packages/storage-plus/src/snapshot.rs

```diff
@@ -0,0 +1,371 @@
+use serde::{Deserialize, Serialize};
+
+use cosmwasm_std::{Order, StdError, StdResult, Storage};
+
+use crate::keys::{EmptyPrefix, PrimaryKey, U64Key};
+use crate::map::Map;
+use crate::path::Path;
+use crate::prefix::Prefix;
+use crate::{Bound, Prefixer};
+
+/// Map that maintains a snapshots of one or more checkpoints.
+/// We can query historical data as well as current state.
+/// What data is snapshotted depends on the Strategy.
+pub struct SnapshotMap<'a, K, T> {
+    primary: Map<'a, K, T>,
+
+    // maps height to number of checkpoints (only used for selected)
+    checkpoints: Map<'a, U64Key, u32>,
+
+    // this stores all changes (key, height). Must differentiate between no data written,
+    // and explicit None (just inserted)
+    changelog: Map<'a, (K, U64Key), ChangeSet<T>>,
+
+    // How aggressive we are about checkpointing all data
+    strategy: Strategy,
+}
+
+#[derive(Clone, Copy, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Strategy {
+    EveryBlock,
+    Never,
+    // Only writes for linked blocks - does a few more reads to save some writes.
+    // Probably uses more gas, but less total disk usage
+    Selected,
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T> {
+    /// Usage: SnapshotMap::new(snapshot_names!("foobar"), Strategy::EveryBlock)
+    pub const fn new(namespaces: SnapshotNamespaces<'a>, strategy: Strategy) -> Self {
+        SnapshotMap {
+            primary: Map::new(namespaces.pk),
+            checkpoints: Map::new(namespaces.checkpoints),
+            changelog: Map::new(namespaces.changelog),
+            strategy,
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.checkpoints
+            .update::<_, StdError>(store, height.into(), |count| {
+                Ok(count.unwrap_or_default() + 1)
+            })?;
+        Ok(())
+    }
+
+    pub fn remove_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        let count = self
+            .checkpoints
+            .may_load(store, height.into())?
+            .unwrap_or_default();
+        if count <= 1 {
+            self.checkpoints.remove(store, height.into());
+            Ok(())
+        } else {
+            self.checkpoints.save(store, height.into(), &(count - 1))
+        }
+    }
+}
+
+impl<'a, K, T> SnapshotMap<'a, K, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a> + Prefixer<'a>,
+{
+    pub fn key(&self, k: K) -> Path<T> {
+        self.primary.key(k)
+    }
+
+    pub fn prefix(&self, p: K::Prefix) -> Prefix<T> {
+        self.primary.prefix(p)
+    }
+
+    /// should_checkpoint looks at the strategy and determines if we want to checkpoint
+    fn should_checkpoint(&self, store: &dyn Storage, k: &K) -> StdResult<bool> {
+        match self.strategy {
+            Strategy::EveryBlock => Ok(true),
+            Strategy::Never => Ok(false),
+            Strategy::Selected => {
```

Comment:
Interesting. A little bit tricky / delicate to use, but can be useful to save some bytes.